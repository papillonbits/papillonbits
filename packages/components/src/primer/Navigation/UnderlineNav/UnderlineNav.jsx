import { Fragment } from 'react'
import cx from 'classnames'
import { v1 as uuidv1 } from 'uuid'
import { getIndexItems, getIndexItemsWithSelected } from '@papillonbits/library/array'
import { useState } from '@papillonbits/library/hooks'
import { propTypes, defaultProps, underlineNavAlignmentType } from './UnderlineNav.prop'
import styles from './UnderlineNav.scss'
import containerStyles from '../../Layout/container.scss'
import { Icon, iconSize } from '../../Icon'

export function UnderlineNav({
  className = defaultProps.className,
  ariaAttr,
  items,
  itemType,
  actions = defaultProps.actions,
  align = defaultProps.align,
  fullContainer = defaultProps.fullContainer,
  onClick,
}) {
  const [indexItems, setIndexItems] = useState(getIndexItems(items))
  const { label, selected, current } = ariaAttr

  function renderActions() {
    return <div className={styles['UnderlineNav-actions']}>{actions}</div>
  }

  function renderBody() {
    return (
      <Fragment>
        {align === underlineNavAlignmentType.right && actions && renderActions()}
        <div className={styles['UnderlineNav-body']} role="tablist">
          {indexItems.map((indexItem) => {
            const { href, icon, text, html, isSelected } = indexItem
            const anchorOtherProps = isSelected ? { ...{ 'aria-selected': selected } } : null
            const buttonOtherProps = isSelected ? { ...{ 'aria-current': current } } : null

            switch (itemType) {
              case 'button':
                return (
                  <button
                    key={uuidv1()}
                    onClick={(e) => {
                      e.preventDefault()
                      const newIndexItems = getIndexItemsWithSelected(indexItems, indexItem)
                      setIndexItems(newIndexItems)
                      onClick(newIndexItems)
                    }}
                    className={styles['UnderlineNav-item']}
                    role="tab"
                    type="button"
                    href={href}
                    {...anchorOtherProps}
                  >
                    {icon && <Icon className={styles['UnderlineNav-octicon']} icon={icon} size={iconSize.small} />}
                    {html || null}
                    {text}
                  </button>
                )
              case 'a':
                return (
                  <a
                    key={uuidv1()}
                    onClick={(e) => {
                      e.preventDefault()
                      const newIndexItems = getIndexItemsWithSelected(indexItems, indexItem)
                      setIndexItems(newIndexItems)
                      onClick(newIndexItems)
                    }}
                    className={styles['UnderlineNav-item']}
                    href={href}
                    {...buttonOtherProps}
                  >
                    {icon && <Icon className={styles['UnderlineNav-octicon']} icon={icon} size={iconSize.small} />}
                    {html || null}
                    {text}
                  </a>
                )
              default:
                return null
            }
          })}
        </div>
        {align === underlineNavAlignmentType.left && actions && renderActions()}
      </Fragment>
    )
  }

  function renderBodyFullContainer() {
    return <div className={cx(containerStyles['container-lg'], styles['UnderlineNav-container'])}>{renderBody()}</div>
  }

  return (
    <nav
      className={cx(className, styles.UnderlineNav, {
        [styles['UnderlineNav--right']]: align === underlineNavAlignmentType.right,
        [styles['UnderlineNav--full']]: fullContainer === true,
      })}
      aria-label={label}
    >
      {fullContainer === false && renderBody()}
      {fullContainer === true && renderBodyFullContainer()}
    </nav>
  )
}

UnderlineNav.propTypes = propTypes
