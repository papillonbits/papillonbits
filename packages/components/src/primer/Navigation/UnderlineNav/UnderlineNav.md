# UnderlineNav

> Use UnderlineNav to style navigation with a minimal underlined selected state, typically used for navigation placed at the top of the page. This component comes with variations to accommodate icons, containers and other content.

# Variants

> - Default
> - WithActions, RightAligned, RightAlignedWithActions
> - WithIcons, FullContainer

## Default

```jsx
<UnderlineNav ariaAttr={ariaAttr} items={buttonItems} itemType={underlineNavItemType.button} onClick={action('onClick')} />
```

## WithActions

```jsx
<UnderlineNav ariaAttr={ariaAttr} items={anchorItems} itemType={underlineNavItemType.a} actions={actions} onClick={action('onClick')} />
```

## RightAligned

```jsx
<UnderlineNav
  ariaAttr={ariaAttr}
  items={anchorItems}
  itemType={underlineNavItemType.a}
  align={underlineNavAlignmentType.right}
  onClick={action('onClick')}
/>
```

## RightAlignedWithActions

```jsx
<UnderlineNav
  ariaAttr={ariaAttr}
  items={anchorItems}
  itemType={underlineNavItemType.a}
  align={underlineNavAlignmentType.right}
  actions={actions}
  onClick={action('onClick')}
/>
```

## WithIcons

```jsx
<UnderlineNav ariaAttr={ariaAttr} items={iconItems} itemType={underlineNavItemType.a} onClick={action('onClick')} />
```

## FullContainer

```jsx
<UnderlineNav
  ariaAttr={ariaAttr}
  items={fullContainerItems}
  itemType={underlineNavItemType.a}
  actions={actions}
  fullContainer
  onClick={action('onClick')}
/>
```
