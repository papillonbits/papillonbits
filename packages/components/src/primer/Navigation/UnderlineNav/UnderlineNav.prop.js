import { string, node, arrayOf, shape, bool, oneOf, func } from 'prop-types'

export const underlineNavItemType = {
  a: 'a',
  button: 'button',
}

export const underlineNavAlignmentType = {
  right: 'right',
  left: 'left',
}

export const propTypes = {
  className: string,
  ariaAttr: shape({
    label: string,
    selected: string,
    current: string,
  }).isRequired,
  items: arrayOf(
    shape({
      href: string,
      icon: string,
      text: string,
      html: node,
      isSelected: bool,
    }),
  ).isRequired,
  itemType: oneOf(Object.keys(underlineNavItemType)).isRequired,
  actions: arrayOf(node),
  align: oneOf(Object.keys(underlineNavAlignmentType)).isRequired,
  fullContainer: bool,
  onClick: func.isRequired,
}

export const defaultProps = {
  className: null,
  actions: null,
  align: 'left',
  fullContainer: false,
}
