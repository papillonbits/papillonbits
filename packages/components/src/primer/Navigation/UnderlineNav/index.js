export { UnderlineNav } from './UnderlineNav'
export { underlineNavItemType, underlineNavAlignmentType } from './UnderlineNav.prop'
