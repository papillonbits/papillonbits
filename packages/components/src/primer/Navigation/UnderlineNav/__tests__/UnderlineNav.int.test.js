import {
  buttonItems,
  anchorItems,
  custom,
  regular,
  withActions,
  rightAligned,
  rightAlignedWithActions,
  withIcons,
  fullContainer,
} from './UnderlineNav.int.story'
import { underlineNavItemType } from '../UnderlineNav.prop'

jest.mock('@papillonbits/library/array', () => ({
  getIndexItems: () => {},
  getIndexItemsWithSelected: () => {},
}))

jest.mock('@papillonbits/library/hooks', () => ({
  useState: () => {},
}))

const libraryArrayMockObject = require('@papillonbits/library/array')
const libraryHooksMockObject = require('@papillonbits/library/hooks')

describe('<UnderlineNav />', () => {
  let indexButtonItemsDataObject = [
    { ...buttonItems[0], index: 0 },
    { ...buttonItems[1], index: 1 },
    { ...buttonItems[2], index: 2 },
    { ...buttonItems[3], index: 3 },
  ]

  let indexAnchorItemsDataObject = [
    { ...anchorItems[0], index: 0 },
    { ...anchorItems[1], index: 1 },
    { ...anchorItems[2], index: 2 },
    { ...anchorItems[3], index: 3 },
  ]

  const newIndexButtonItemsWithSelectedDataObject = [
    { ...buttonItems[0], index: 0, isSelected: true },
    { ...buttonItems[1], index: 1, isSelected: false },
    { ...buttonItems[2], index: 2, isSelected: false },
    { ...buttonItems[3], index: 3, isSelected: false },
  ]

  const newIndexAnchorItemsWithSelectedDataObject = [
    { ...anchorItems[0], index: 0, isSelected: true },
    { ...anchorItems[1], index: 1, isSelected: false },
    { ...anchorItems[2], index: 2, isSelected: false },
    { ...anchorItems[3], index: 3, isSelected: false },
  ]

  const onClickMockFn = jest.fn()

  const setIndexButtonItemsMockFn = jest.fn((indexButtonItems) => {
    indexButtonItemsDataObject = indexButtonItems
  })

  const setIndexAnchorItemsMockFn = jest.fn((indexAnchorItems) => {
    indexAnchorItemsDataObject = indexAnchorItems
  })

  function mockLibraryForButtonItems() {
    jest.spyOn(libraryArrayMockObject, 'getIndexItems').mockReturnValue(indexButtonItemsDataObject)

    jest.spyOn(libraryArrayMockObject, 'getIndexItemsWithSelected').mockReturnValue(newIndexButtonItemsWithSelectedDataObject)

    jest.spyOn(libraryHooksMockObject, 'useState').mockImplementation(() => [indexButtonItemsDataObject, setIndexButtonItemsMockFn])
  }

  function mockLibraryForAnchorItems() {
    jest.spyOn(libraryArrayMockObject, 'getIndexItems').mockReturnValue(indexAnchorItemsDataObject)

    jest.spyOn(libraryArrayMockObject, 'getIndexItemsWithSelected').mockReturnValue(newIndexAnchorItemsWithSelectedDataObject)

    jest.spyOn(libraryHooksMockObject, 'useState').mockImplementation(() => [indexAnchorItemsDataObject, setIndexAnchorItemsMockFn])
  }

  describe('Event', () => {
    test('must return new items without index when button items and onClick()', () => {
      mockLibraryForButtonItems()
      const mountCustomRender = global.renderMount(custom(buttonItems, underlineNavItemType.button, onClickMockFn))
      mountCustomRender.find('button').first().simulate('click')
      expect(mountCustomRender.props().onClick).toBe(onClickMockFn)
      expect(mountCustomRender.props().onClick).toHaveBeenCalledWith(newIndexButtonItemsWithSelectedDataObject)
    })

    test('must return new items without index when anchor items and onClick()', () => {
      mockLibraryForAnchorItems()
      const mountCustomRender = global.renderMount(custom(anchorItems, underlineNavItemType.a, onClickMockFn))
      mountCustomRender.find('a').first().simulate('click')
      expect(mountCustomRender.props().onClick).toBe(onClickMockFn)
      expect(mountCustomRender.props().onClick).toHaveBeenCalledWith(newIndexAnchorItemsWithSelectedDataObject)
    })
  })

  describe('Render', () => {
    test('must match regular()', () => {
      mockLibraryForButtonItems()
      expect(global.renderToJSON(regular())).toMatchSnapshot()
    })

    test('must match withActions()', () => {
      mockLibraryForButtonItems()
      expect(global.renderToJSON(withActions())).toMatchSnapshot()
    })

    test('must match rightAligned()', () => {
      mockLibraryForButtonItems()
      expect(global.renderToJSON(rightAligned())).toMatchSnapshot()
    })

    test('must match rightAlignedWithActions()', () => {
      mockLibraryForButtonItems()
      expect(global.renderToJSON(rightAlignedWithActions())).toMatchSnapshot()
    })

    test('must match withIcons()', () => {
      mockLibraryForButtonItems()
      expect(global.renderToJSON(withIcons())).toMatchSnapshot()
    })

    test('must match fullContainer()', () => {
      mockLibraryForButtonItems()
      expect(global.renderToJSON(fullContainer())).toMatchSnapshot()
    })
  })

  describe('State', () => {
    test('must reset new items with index when button items and onClick()', () => {
      mockLibraryForButtonItems()
      const mountCustomRender = global.renderMount(custom(buttonItems, underlineNavItemType.button, onClickMockFn))
      mountCustomRender.find('button').first().simulate('click')
      expect(setIndexButtonItemsMockFn).toHaveBeenCalledWith(newIndexButtonItemsWithSelectedDataObject)
      expect(indexButtonItemsDataObject).toBe(newIndexButtonItemsWithSelectedDataObject)
    })

    test('must reset new items with index when anchor items and onClick()', () => {
      mockLibraryForAnchorItems()
      const mountCustomRender = global.renderMount(custom(anchorItems, underlineNavItemType.a, onClickMockFn))
      mountCustomRender.find('a').first().simulate('click')
      expect(setIndexAnchorItemsMockFn).toHaveBeenCalledWith(newIndexAnchorItemsWithSelectedDataObject)
      expect(indexAnchorItemsDataObject).toBe(newIndexAnchorItemsWithSelectedDataObject)
    })
  })
})
