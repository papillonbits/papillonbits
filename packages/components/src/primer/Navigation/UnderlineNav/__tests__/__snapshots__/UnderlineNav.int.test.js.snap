// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UnderlineNav /> Render must match fullContainer() 1`] = `
<nav
  aria-label="Foo bar"
  className="UnderlineNav UnderlineNav--full"
>
  <div
    className="container-lg UnderlineNav-container"
  >
    <div
      className="UnderlineNav-body"
      role="tablist"
    >
      <a
        aria-current="page"
        className="UnderlineNav-item"
        onClick={[Function]}
      >
        Item 1
      </a>
      <a
        className="UnderlineNav-item"
        onClick={[Function]}
      >
        Item 2
      </a>
      <a
        className="UnderlineNav-item"
        onClick={[Function]}
      >
        Item 3
      </a>
      <a
        className="UnderlineNav-item"
        onClick={[Function]}
      >
        Item 4
      </a>
    </div>
    <div
      className="UnderlineNav-actions"
    >
      <a
        className="btn"
        href="#url"
      >
        Button
      </a>
    </div>
  </div>
</nav>
`;

exports[`<UnderlineNav /> Render must match regular() 1`] = `
<nav
  aria-label="Foo bar"
  className="UnderlineNav"
>
  <div
    className="UnderlineNav-body"
    role="tablist"
  >
    <button
      aria-selected="true"
      className="UnderlineNav-item"
      onClick={[Function]}
      role="tab"
      type="button"
    >
      Item 1
    </button>
    <button
      className="UnderlineNav-item"
      onClick={[Function]}
      role="tab"
      type="button"
    >
      Item 2
    </button>
    <button
      className="UnderlineNav-item"
      onClick={[Function]}
      role="tab"
      type="button"
    >
      Item 3
    </button>
    <button
      className="UnderlineNav-item"
      onClick={[Function]}
      role="tab"
      type="button"
    >
      Item 4
    </button>
  </div>
</nav>
`;

exports[`<UnderlineNav /> Render must match rightAligned() 1`] = `
<nav
  aria-label="Foo bar"
  className="UnderlineNav UnderlineNav--right"
>
  <div
    className="UnderlineNav-body"
    role="tablist"
  >
    <a
      aria-current="page"
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 1
    </a>
    <a
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 2
    </a>
    <a
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 3
    </a>
    <a
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 4
    </a>
  </div>
</nav>
`;

exports[`<UnderlineNav /> Render must match rightAlignedWithActions() 1`] = `
<nav
  aria-label="Foo bar"
  className="UnderlineNav UnderlineNav--right"
>
  <div
    className="UnderlineNav-actions"
  >
    <a
      className="btn"
      href="#url"
    >
      Button
    </a>
  </div>
  <div
    className="UnderlineNav-body"
    role="tablist"
  >
    <a
      aria-current="page"
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 1
    </a>
    <a
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 2
    </a>
    <a
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 3
    </a>
    <a
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 4
    </a>
  </div>
</nav>
`;

exports[`<UnderlineNav /> Render must match withActions() 1`] = `
<nav
  aria-label="Foo bar"
  className="UnderlineNav"
>
  <div
    className="UnderlineNav-body"
    role="tablist"
  >
    <a
      aria-current="page"
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 1
    </a>
    <a
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 2
    </a>
    <a
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 3
    </a>
    <a
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 4
    </a>
  </div>
  <div
    className="UnderlineNav-actions"
  >
    <a
      className="btn"
      href="#url"
    >
      Button
    </a>
  </div>
</nav>
`;

exports[`<UnderlineNav /> Render must match withIcons() 1`] = `
<nav
  aria-label="Foo bar"
  className="UnderlineNav"
>
  <div
    className="UnderlineNav-body"
    role="tablist"
  >
    <a
      aria-current="page"
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 1
    </a>
    <a
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 2
    </a>
    <a
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 3
    </a>
    <a
      className="UnderlineNav-item"
      onClick={[Function]}
    >
      Item 4
    </a>
  </div>
</nav>
`;
